// @zowe/cli/lib/zostso/src/cli/ZosTso.definition.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
Object.defineProperty(exports, "__esModule", { value: true });
const Start_definition_1 = require("./start/Start.definition");
const Send_definition_1 = require("./send/Send.definition");
const Stop_definition_1 = require("./stop/Stop.definition");
const Ping_definition_1 = require("./ping/Ping.definition");
const Issue_definition_1 = require("./issue/Issue.definition");
const zosmf_1 = require("../../../zosmf");
exports.definition = {
    name: "zos-tso",
    aliases: ["tso"],
    type: "group",
    summary: "Interact with TSO",
    description: "Issue TSO commands and interact with TSO address spaces",
    children: [
        Send_definition_1.SendCommand, Start_definition_1.StartCommand, Ping_definition_1.PingCommand, Stop_definition_1.StopCommand, Issue_definition_1.IssueCommand
    ],
    passOn: [{
            property: "options",
            value: zosmf_1.ZosmfSession.ZOSMF_CONNECTION_OPTIONS,
            merge: true,
            ignoreNodes: [
                {
                    type: "group"
                }
            ]
        }]
};
module.exports = exports.definition;
