// @zowe/cli/lib/zostso/src/api/TsoConstants.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
Object.defineProperty(exports, "__esModule", { value: true });
const imperative_1 = require("@zowe/imperative");
/**
 * Constants for TSO related info
 * @export
 * @class TsoConstants
 */
class TsoConstants {
}
/**
 * Quert id of logonProcedure passed to z/OSMF URI
 * @static
 * @type {string}
 * @memberof TsoConstants
 */
TsoConstants.PARM_PROC = "proc";
/**
 * Quert id of character-set passed to z/OSMF URI
 * @static
 * @type {string}
 * @memberof TsoConstants
 */
TsoConstants.PARM_CHSET = "chset";
/**
 * Quert id of code page passed to z/OSMF URI
 * @static
 * @type {string}
 * @memberof TsoConstants
 */
TsoConstants.PARM_CPAGE = "cpage";
/**
 * Quert id of rows passed to z/OSMF URI
 * @static
 * @type {string}
 * @memberof TsoConstants
 */
TsoConstants.PARM_ROWS = "rows";
/**
 * Quert id of columns passed to z/OSMF URI
 * @static
 * @type {string}
 * @memberof TsoConstants
 */
TsoConstants.PARM_COLS = "cols";
/**
 * Quert id of account number passed to z/OSMF URI
 * @static
 * @type {string}
 * @memberof TsoConstants
 */
TsoConstants.PARM_ACCT = "acct";
/**
 * Quert id of region size passed to z/OSMF URI
 * @static
 * @type {string}
 * @memberof TsoConstants
 */
TsoConstants.PARM_RSIZE = "rsize";
/**
 * Default character-set value
 * @static
 * @type {string}
 * @memberof TsoConstants
 */
TsoConstants.DEFAULT_CHSET = "697";
/**
 * Default code page value
 * @static
 * @type {string}
 * @memberof TsoConstants
 */
TsoConstants.DEFAULT_CPAGE = "1047";
/**
 * Default number of rows value
 * @static
 * @type {string}
 * @memberof TsoConstants
 */
TsoConstants.DEFAULT_ROWS = "24";
/**
 * Default number of columns value
 * @static
 * @type {string}
 * @memberof TsoConstants
 */
TsoConstants.DEFAULT_COLS = "80";
/**
 * Default region-size value
 * @static
 * @type {string}
 * @memberof TsoConstants
 */
TsoConstants.DEFAULT_RSIZE = "4096";
/**
 * Default logonProcedure value
 * @static
 * @type {string}
 * @memberof TsoConstants
 */
TsoConstants.DEFAULT_PROC = "IZUFPROC";
/**
 * URI base for TSO API
 * @static
 * @type {string}
 * @memberof TsoConstants
 */
TsoConstants.RESOURCE = "/zosmf/tsoApp";
/**
 * URI for starting TSO
 * @static
 * @type {string}
 * @memberof TsoConstants
 */
TsoConstants.RES_START_TSO = "tso";
/**
 * Param for not reading reply
 * @static
 * @type {string}
 * @memberof TsoConstants
 */
TsoConstants.RES_DONT_READ_REPLY = "?readReply=false";
/**
 * URI for TSO Ping API
 * @static
 * @type {string}
 * @memberof TsoConstants
 */
TsoConstants.RES_PING = TsoConstants.RESOURCE + "/" + TsoConstants.RES_START_TSO + "/ping";
/**
 * Tso response message type - prompt
 * @static
 * @type {string}
 * @memberof TsoConstants
 */
TsoConstants.TSO_PROMPT = "TSO PROMPT";
/**
 * Tso response message type - message
 * @static
 * @type {string}
 * @memberof TsoConstants
 */
TsoConstants.TSO_MESSAGE = "TSO MESSAGE";
exports.TsoConstants = TsoConstants;
/**
 * No Session provided error message
 * @static
 * @type {IMessageDefinition}
 * @memberof TsoConstants
 */
exports.noSessionTso = {
    message: imperative_1.apiErrorHeader + `No session was supplied.`
};
/**
 * No input parameters were provided error message
 * @static
 * @type {IMessageDefinition}
 * @memberof TsoConstants
 */
exports.noTsoStartInput = {
    message: imperative_1.apiErrorHeader + `No tso start address space parameters were supplied.`
};
/**
 * No input parameters were provided error message
 * @static
 * @type {IMessageDefinition}
 * @memberof TsoConstants
 */
exports.noTsoIssueInput = {
    message: imperative_1.apiErrorHeader + `No tso issue command parameters were supplied.`
};
/**
 * No input parameters for stop were provided error message
 * @static
 * @type {IMessageDefinition}
 * @memberof TsoConstants
 */
exports.noTsoStopInput = {
    message: imperative_1.apiErrorHeader + `No tso stop address space parameters were supplied.`
};
/**
 * No ZOSMF response was received error message
 * @static
 * @type {IMessageDefinition}
 * @memberof TsoConstants
 */
exports.noZosmfResponse = {
    message: imperative_1.apiErrorHeader + `No z/OSMF response was received.`
};
/**
 * No servlet key was provided for Ping command error message
 * @static
 * @type {IMessageDefinition}
 * @memberof TsoConstants
 */
exports.noPingInput = {
    message: imperative_1.apiErrorHeader + `No servlet supplied.`
};
/**
 * No account number was provided error message
 * @static
 * @type {IMessageDefinition}
 * @memberof TsoConstants
 */
exports.noAccountNumber = {
    message: imperative_1.apiErrorHeader + `No account number was supplied.`
};
/**
 * No servlet key was provided error message
 * @static
 * @type {IMessageDefinition}
 * @memberOf TsoConstants
 */
exports.noServletKeyInput = {
    message: imperative_1.apiErrorHeader + `No servlet key was supplied.`
};
/**
 * No data parameter string was supplied error message
 * @static
 * @type {IMessageDefinition}
 * @memberof TsoConstants
 */
exports.noDataInput = {
    message: imperative_1.apiErrorHeader + `No data parameter string was supplied.`
};
/**
 * No command text was provided error message
 * @static
 * @type {IMessageDefinition}
 * @memberof TsoConstants
 */
exports.noCommandInput = {
    message: imperative_1.apiErrorHeader + `No command text was provided.`
};
