// @zowe/imperative/lib/utilities/src/ProcessUtils.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * This enum represents the possible results from isGuiAvailable.
 */
var GuiResult;
(function (GuiResult) {
    /** A GUI is available */
    GuiResult[GuiResult["GUI_AVAILABLE"] = 0] = "GUI_AVAILABLE";
    /** No GUI because this is an SSH connection. */
    GuiResult[GuiResult["NO_GUI_SSH"] = 1] = "NO_GUI_SSH";
    /** No GUI because The $DISPLAY environment variable is not set. */
    GuiResult[GuiResult["NO_GUI_NO_DISPLAY"] = 2] = "NO_GUI_NO_DISPLAY";
})(GuiResult = exports.GuiResult || (exports.GuiResult = {}));
/**
 * A collection of utilities related to the running process.
 * @export
 * @class ProcessUtils
 */
class ProcessUtils {
    // __________________________________________________________________________
    /**
     * Process utility to wrap callback process routines into promises
     * Turn nextTick into a promise to prevent nesting
     * @static
     * @param {() => void} callback - called before promise is resolved
     * @param {...any[]} args - arguments passed to the callback
     * @returns {Promise<void>} - fullfilled whenever callback is invoked
     * @memberof ProcessUtils
     */
    static nextTick(callback, ...args) {
        return new Promise((resolve, reject) => {
            process.nextTick(() => {
                callback(...args);
                resolve();
            });
        });
    }
    // __________________________________________________________________________
    /**
     * Is a Graphical User Interface avaiable in the environment in which
     * the current command is running?
     *
     * @returns {boolean} - True if GUI. False when no GUI.
     */
    static isGuiAvailable() {
        /* If any of the SSH environment variables are defined,
         * then we are in an ssh session --> no GUI.
         */
        if (typeof process.env.SSH_CONNECTION !== "undefined" ||
            typeof process.env.SSH_CLIENT !== "undefined" ||
            typeof process.env.SSH_TTY !== "undefined") {
            return GuiResult.NO_GUI_SSH;
        }
        /* On linux (and MAC) the DISPLAY environment variable
         * indicates that we are in an X-Window environment.
         */
        if (process.platform !== "win32") {
            if (typeof process.env.DISPLAY === "undefined" ||
                process.env.DISPLAY === "") {
                return GuiResult.NO_GUI_NO_DISPLAY;
            }
        }
        // otherwise we assume we have a GUI
        return GuiResult.GUI_AVAILABLE;
    }
}
exports.ProcessUtils = ProcessUtils;
