// @zowe/cli/lib/zosfiles/src/cli/create/cPds/CPDS.definition.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
Object.defineProperty(exports, "__esModule", { value: true });
const create_1 = require("../../../api/methods/create");
const Create_options_1 = require("../Create.options");
// Does not use the import in anticipation of some internationalization work to be done later.
const strings = require("../../-strings-/en").default.CREATE;
/**
 * Create C-code PDS command definition containing its description, examples and/or options
 * @type {ICommandDefinition}
 */
exports.CPDSDefinition = {
    name: "data-set-c",
    aliases: ["dsc"],
    description: strings.ACTIONS.DATA_SET_C.DESCRIPTION,
    type: "command",
    handler: __dirname + "/CPDS.handler",
    profile: {
        optional: ["zosmf"],
    },
    positionals: [
        {
            name: "dataSetName",
            type: "string",
            description: strings.POSITIONALS.DATASETNAME,
            required: true,
        },
    ],
    options: [
        Object.assign({}, Create_options_1.ZosFilesCreateExtraOptions.size, { defaultValue: create_1.CreateDefaults.DATA_SET.C.primary + create_1.CreateDefaults.DATA_SET.C.alcunit }),
        Create_options_1.ZosFilesCreateOptions.volser,
        Object.assign({}, Create_options_1.ZosFilesCreateOptions.secondary, { defaultValue: create_1.CreateDefaults.DATA_SET.C.secondary }),
        Object.assign({}, Create_options_1.ZosFilesCreateOptions.dirblk, { defaultValue: create_1.CreateDefaults.DATA_SET.C.dirblk }),
        Object.assign({}, Create_options_1.ZosFilesCreateOptions.recfm, { defaultValue: create_1.CreateDefaults.DATA_SET.C.recfm }),
        Object.assign({}, Create_options_1.ZosFilesCreateOptions.blksize, { defaultValue: create_1.CreateDefaults.DATA_SET.C.blksize }),
        Object.assign({}, Create_options_1.ZosFilesCreateOptions.lrecl, { defaultValue: create_1.CreateDefaults.DATA_SET.C.lrecl }),
        Create_options_1.ZosFilesCreateOptions.storeclass,
        Create_options_1.ZosFilesCreateOptions.mgntclass,
        Create_options_1.ZosFilesCreateOptions.dataclass,
        Create_options_1.ZosFilesCreateOptions.unit,
        Create_options_1.ZosFilesCreateOptions.dsntype,
        Create_options_1.ZosFilesCreateExtraOptions.showAttributes,
    ].sort((a, b) => a.name.localeCompare(b.name)),
    examples: [
        {
            description: strings.ACTIONS.DATA_SET_C.EXAMPLES.EX1,
            options: "NEW.CCODE.DATASET",
        }
    ]
};
