// @zowe/cli/lib/zosuss/src/cli/ZosUss.definition.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
Object.defineProperty(exports, "__esModule", { value: true });
const Issue_definition_1 = require("./issue/Issue.definition");
const SshSession_1 = require("../SshSession");
exports.definition = {
    name: "zos-uss",
    aliases: ["uss"],
    type: "group",
    summary: "Issue z/OS USS commands and receive responses",
    description: "Issue z/OS USS commands remotely using an SSH session. Output from the commands is displayed on the local terminal.",
    children: [
        Issue_definition_1.IssueCommand,
    ],
    passOn: [
        {
            property: "options",
            value: SshSession_1.SshSession.SSH_CONNECTION_OPTIONS,
            merge: true,
            ignoreNodes: [
                { type: "group" }
            ]
        }
    ]
};
module.exports = exports.definition;
