// @zowe/cli/lib/workflows/src/cli/Workflows.definition.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
const Delete_definition_1 = require("./delete/Delete.definition");
const Create_definition_1 = require("./create/Create.definition");
const Start_definition_1 = require("./start/Start.definition");
const zosmf_1 = require("../../../zosmf");
const List_definition_1 = require("./list/List.definition");
const Archive_definition_1 = require("./archive/Archive.definition");
/**
 * This object defines the top level command group for zosworkflows. This is not
 * something that is intended to be used outside of this npm package.
 *
 * @private
 */
const definition = {
    name: "zos-workflows",
    aliases: ["wf"],
    type: "group",
    summary: "Create and manage z/OSMF workflows",
    description: "Create and manage z/OSMF workflows on a z/OS system",
    children: [
        Create_definition_1.CreateDefinition,
        Start_definition_1.StartDefinition,
        List_definition_1.ListDefinition,
        Archive_definition_1.ArchiveDefinition,
        Delete_definition_1.DeleteDefinition,
    ],
    passOn: [
        {
            property: "options",
            value: zosmf_1.ZosmfSession.ZOSMF_CONNECTION_OPTIONS,
            merge: true,
            ignoreNodes: [
                { type: "group" }
            ]
        }
    ]
};
module.exports = definition;
