// @zowe/cli/lib/provisioning/src/cli/Provisioning.definition.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
const List_definition_1 = require("./list/List.definition");
const Provision_definition_1 = require("./provision/Provision.definition");
const Perform_definition_1 = require("./perform/Perform.definition");
const Delete_definition_1 = require("./delete/Delete.definition");
const zosmf_1 = require("../../../zosmf");
const definition = {
    name: "provisioning",
    aliases: ["pv"],
    type: "group",
    summary: "Perform z/OSMF provisioning tasks",
    description: "Perform z/OSMF provisioning tasks on Published Templates in the Service Catalog " +
        "and Provisioned Instances in the Service Registry.",
    children: [
        List_definition_1.ListCommand, Provision_definition_1.ProvisionCommand, Perform_definition_1.PerformCommand, Delete_definition_1.DeleteCommand
    ],
    passOn: [
        {
            property: "options",
            value: zosmf_1.ZosmfSession.ZOSMF_CONNECTION_OPTIONS,
            merge: true,
            ignoreNodes: [
                { type: "group" }
            ]
        }
    ]
};
module.exports = definition;
