// @zowe/imperative/lib/cmd/src/profiles/CommandProfiles.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
Object.defineProperty(exports, "__esModule", { value: true });
const error_1 = require("../../../error");
const expect_1 = require("../../../expect");
/**
 * Profiles map created by the command profile loader and passed to the handler via parameters. Handlers can
 * retreive loaded profiles from the map via the profile type.
 * @export
 * @class CommandProfiles
 */
class CommandProfiles {
    /**
     * Creates an instance of CommandProfiles.
     * @param {Map<string, IProfile[]>} map - The map of profiles
     * @memberof CommandProfiles
     */
    constructor(map) {
        // Simple validation of input parameters
        const err = "Command Profiles Internal Error:";
        expect_1.ImperativeExpect.toNotBeNullOrUndefined(map, `${err} No map was supplied.`);
        expect_1.ImperativeExpect.toBeEqual(map instanceof Map, true, `${err} The "map" supplied is not an instance of a map.`);
        // Ensure the correctness of each map entry
        map.forEach((value, key) => {
            expect_1.ImperativeExpect.toBeAnArray(value, `${err} The "profiles" supplied for type "${key}" is NOT an array.`);
            expect_1.ImperativeExpect.toBeEqual((value.length > 0), true, `${err} No profiles supplied for type "${key}".`);
        });
        this.mMap = map;
    }
    /**
     * Gets the first (or by name) profile in the map - automatically throws an exception (unless disabled)
     * @template T - The expected profile mapping to be returned
     * @param {string} type - The profile type
     * @param {string} [name=""] - The name of the profile to retrieve
     * @param {boolean} [failNotFound=true] - Automatically throws an imperative exception if not profiles are not
     * found - this is provided as convince for the handlers (will fail your command if not found) - This would
     * normally be the result of a command configuration problem.
     * @returns {T} - The first profile in the map (or the one located by name)
     * @memberof CommandProfiles
     */
    get(type, failNotFound = true, name = "") {
        let profile;
        // If a profile is returned for the type, then we'll check if a profile of a specific name was requseted
        // if not, just return the first profile found (first loaded)
        if (this.map.get(type) != null) {
            if (name != null && name.trim().length > 0) {
                for (const prof of this.map.get(type)) {
                    if (prof.name === name) {
                        profile = prof;
                        break;
                    }
                }
            }
            else {
                profile = this.map.get(type)[0];
            }
        }
        else if (failNotFound) {
            this.fail(type);
        }
        return profile;
    }
    /**
     * Gets all profiles for the type specified,
     * @template T - extends IProfile
     * @param {string} type - The profile type to retrieve loaded profiles
     * @returns {T[]} - The list of profile types
     * @param {boolean} [failNotFound=true] - Automatically throws an imperative exception if not profiles are not
     * found - this is provided as convince for the handlers (will fail your command if not found) - This would
     * normally be the result of a command configuration problem.
     * @memberof CommandProfiles
     */
    getAll(type, failNotFound = true) {
        const profiles = this.map.get(type);
        if (profiles == null && failNotFound) {
            this.fail(type);
        }
        return profiles;
    }
    /**
     * Internal accessor for the map
     * @readonly
     * @private
     * @type {Map<string, IProfile[]>} - The profile Map
     * @memberof CommandProfiles
     */
    get map() {
        return this.mMap;
    }
    /**
     * Throw an error failing the get(requested by the caller if no profiles are available)
     * @private
     * @param {string} type - the profile type to get from the map
     * @memberof CommandProfiles
     */
    fail(type) {
        throw new error_1.ImperativeError({
            msg: `Internal Error: No profiles of type "${type}" were loaded for this command.`,
            additionalDetails: `This error can occur for one of two reasons:\n` +
                ` - The "profile" property on the command definition document ` +
                `does NOT specify the requested profile type\n` +
                ` - The profile type is marked "optional", no profiles of type "${type}" have been created, ` +
                `and the command handler requested a profile of type "${type}" with "failNotFound=true"`
        });
    }
}
exports.CommandProfiles = CommandProfiles;
