// @zowe/cli/lib/zosconsole/src/api/ConsoleConstants.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
Object.defineProperty(exports, "__esModule", { value: true });
const imperative_1 = require("@zowe/imperative");
class ConsoleConstants {
}
ConsoleConstants.CLASS = "Consoles";
ConsoleConstants.RESOURCE = "/zosmf/restconsoles/consoles";
ConsoleConstants.RES_DEF_CN = "defcn";
ConsoleConstants.SOL_MSGS = "/solmsgs";
ConsoleConstants.DEFAULT_FOLLOWUP_ATTEMPTS = 1;
ConsoleConstants.DEFAULT_TIMEOUT = 0;
exports.ConsoleConstants = ConsoleConstants;
exports.displayResponse = {
    message: `Populating console response: {{data}}`
};
exports.displayError = {
    message: imperative_1.apiErrorHeader + ` Populating console error: {{data}}`
};
exports.displayCollectResponse = {
    message: `Collect response: {{data}}`
};
exports.displayCollectError = {
    message: imperative_1.apiErrorHeader + ` Collect error: {{data}}`
};
exports.decreaseCounter = {
    message: `Decreasing follow up counter`
};
exports.resetCounter = {
    message: `Reset follow up counter`
};
exports.noSession = {
    message: imperative_1.apiErrorHeader + `No session was supplied.`
};
exports.noZosmfInput = {
    message: imperative_1.apiErrorHeader + `No zosmf console input parameters were supplied.`
};
exports.noConsoleInput = {
    message: imperative_1.apiErrorHeader + `No console issue parameters were supplied.`
};
exports.noConsoleName = {
    message: imperative_1.apiErrorHeader + `No console name was supplied.`
};
exports.noCommandKey = {
    message: imperative_1.apiErrorHeader + `No command response key was supplied.`
};
exports.noCollectParameters = {
    message: imperative_1.apiErrorHeader + `No console collect parameters were supplied.`
};
exports.collectProcessingDetails = {
    message: `Collect processing - timeout: {{timer}}, followUpCounter: {{counter}}`
};
