// @zowe/imperative/lib/imperative/src/config/cmd/list/list.definition.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
/**
 * Definition of the list command.
 * @type {ICommandDefinition}
 */
exports.listDefinition = {
    name: "list",
    aliases: ["li"],
    type: "command",
    handler: path_1.join(__dirname, "list.handler"),
    summary: "List all configuration setting options",
    description: "List all configuration setting options.",
    options: [
        {
            name: "values",
            type: "boolean",
            description: "Show values for every option",
        },
    ],
    examples: [
        {
            options: "",
            description: "List all configuration setting options"
        },
        {
            options: "--values",
            description: "List all configuration setting options with values"
        }
    ],
};
