// @zowe/imperative/lib/imperative/src/profiles/builders/ProfilesListCommandBuilder.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
Object.defineProperty(exports, "__esModule", { value: true });
const ProfilesCommandBuilder_1 = require("./ProfilesCommandBuilder");
const constants_1 = require("../../../../constants");
const messages_1 = require("../../../../messages");
const utilities_1 = require("../../../../utilities");
const profiles_1 = require("../../../../profiles");
/**
 * Used to build profile update command definitions.
 * Used automatically if you allow the "profiles" command group to be generated
 */
class ProfilesListCommandBuilder extends ProfilesCommandBuilder_1.ProfilesCommandBuilder {
    /**
     * Construct the builder based on the schema.
     * @param profileType - the profile type to generate the command definition for
     * @param {Logger} logger - logger instance to use for the builder class
     * @param {ICommandProfileTypeConfiguration} profileConfig: Imperative profile configuration for this type of profile
     */
    constructor(profileType, logger, profileConfig) {
        super(profileType, logger, profileConfig);
    }
    /**
     * Gets the "action" that this command builder is building.
     * @return {string}: The "create" action string
     */
    getAction() {
        return constants_1.Constants.LIST_ACTION;
    }
    /**
     * Build the full command - includes action group and object command.
     * @return {ICommandDefinition}: The command definition.
     */
    buildFull() {
        return this.buildProfileSegmentFromSchema();
    }
    /**
     * Builds only the "profile" segment from the profile schema.
     * @return {ICommandDefinition}
     */
    buildProfileSegmentFromSchema() {
        const profileCommand = {
            name: this.mProfileType + "-profiles",
            aliases: [this.mProfileType],
            summary: utilities_1.TextUtils.formatMessage(messages_1.listProfileCommandDesc.message, { type: this.mProfileType }),
            description: this.mSchema.description,
            type: "command",
            handler: __dirname + "/../handlers/ListProfilesHandler",
            customize: {},
            options: [
                {
                    name: "show-contents",
                    aliases: ["sc"],
                    description: utilities_1.TextUtils.formatMessage(messages_1.listProfileVerboseOptionDesc.message, {
                        type: this.mProfileType
                    }),
                    type: "boolean",
                    required: false
                }
            ],
            examples: [
                {
                    options: "",
                    description: utilities_1.TextUtils.formatMessage(messages_1.listProfileExample.message, {
                        type: this.mProfileType
                    }),
                },
                {
                    options: "--sc",
                    description: utilities_1.TextUtils.formatMessage(messages_1.listProfileExampleShowContents.message, {
                        type: this.mProfileType
                    }),
                }
            ]
        };
        profileCommand.customize[profiles_1.ProfilesConstants.PROFILES_COMMAND_TYPE_KEY] = this.mProfileType;
        return profileCommand;
    }
}
exports.ProfilesListCommandBuilder = ProfilesListCommandBuilder;
