// @zowe/imperative/lib/imperative/src/profiles/builders/ProfilesDeleteCommandBuilder.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
Object.defineProperty(exports, "__esModule", { value: true });
const ProfilesCommandBuilder_1 = require("./ProfilesCommandBuilder");
const constants_1 = require("../../../../constants");
const index_1 = require("../../../../messages/index");
const utilities_1 = require("../../../../utilities");
const profiles_1 = require("../../../../profiles");
/**
 * Used to build delete profile command definitions.
 * Used automatically if you allow the "profiles" command group to be generated
 */
class ProfilesDeleteCommandBuilder extends ProfilesCommandBuilder_1.ProfilesCommandBuilder {
    /**
     * Gets the "action" that this command builder is building.
     * @return {string}: The "create" action string
     */
    getAction() {
        return constants_1.Constants.DELETE_ACTION;
    }
    /**
     * Build the full command - includes action group and object command.
     * @return {ICommandDefinition}: The command definition.
     */
    buildFull() {
        return this.buildProfileSegmentFromSchema();
    }
    /**
     * Builds only the "profile" segment from the profile schema.
     * @return {ICommandDefinition}
     */
    buildProfileSegmentFromSchema() {
        const profileCommand = {
            name: this.mProfileType + "-profile",
            aliases: [this.mProfileType],
            summary: utilities_1.TextUtils.formatMessage(index_1.deleteProfileActionDesc.message, { type: this.mProfileType }),
            description: utilities_1.TextUtils.formatMessage(index_1.deleteProfileCommandDesc.message, { type: this.mProfileType }),
            type: "command",
            handler: __dirname + "/../handlers/NewDeleteProfilesHandler",
            customize: {},
            options: [],
            positionals: [{
                    name: constants_1.Constants.PROFILE_NAME_OPTION,
                    description: utilities_1.TextUtils.formatMessage(index_1.deleteProfileNameDesc.message, {
                        type: this.mProfileType,
                        typeOption: profiles_1.ProfileUtils.getProfileOption(this.mProfileType)
                    }),
                    type: "string",
                    required: true,
                }],
            examples: [
                {
                    options: "profilename",
                    description: utilities_1.TextUtils.formatMessage(index_1.deleteProfileExample.message, {
                        type: this.mProfileType,
                        name: "profilename"
                    })
                }
            ]
        };
        profileCommand.customize[profiles_1.ProfilesConstants.PROFILES_COMMAND_TYPE_KEY] = this.mProfileType;
        profileCommand.options.push({
            name: "force",
            aliases: [],
            description: utilities_1.TextUtils.formatMessage(index_1.deleteProfileForceOptionDesc.message, {
                type: this.mProfileType,
                typeOption: profiles_1.ProfileUtils.getProfileOption(this.mProfileType)
            }),
            type: "boolean",
            required: false,
        });
        return profileCommand;
    }
}
exports.ProfilesDeleteCommandBuilder = ProfilesDeleteCommandBuilder;
