// @zowe/imperative/lib/imperative/src/profiles/builders/ProfilesShowDependenciesCommandBuilder.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
Object.defineProperty(exports, "__esModule", { value: true });
const ProfilesCommandBuilder_1 = require("./ProfilesCommandBuilder");
const constants_1 = require("../../../../constants");
const messages_1 = require("../../../../messages");
const utilities_1 = require("../../../../utilities");
const profiles_1 = require("../../../../profiles");
/**
 * Used to build profile create command definitions.
 * Used automatically if you allow the "profiles" command group to be generated
 */
class ProfilesShowDependenciesCommandBuilder extends ProfilesCommandBuilder_1.ProfilesCommandBuilder {
    /**
     * Gets the "action" that this command builder is building.
     * @return {string}: The "create" action string
     */
    getAction() {
        return constants_1.Constants.SHOW_DEPS_ACTION;
    }
    /**
     * Build the full command - includes action group and object command.
     * @return {ICommandDefinition}: The command definition.
     */
    buildFull() {
        return this.buildProfileSegmentFromSchema();
    }
    /**
     * Builds only the "profile" segment from the profile schema.
     * @return {ICommandDefinition}
     */
    buildProfileSegmentFromSchema() {
        const profileCommand = {
            name: this.mProfileType + "-profile",
            aliases: [this.mProfileType],
            summary: utilities_1.TextUtils.formatMessage(messages_1.showDependenciesCommandDesc.message, { type: this.mProfileType }),
            description: this.mSchema.description,
            type: "command",
            handler: __dirname + "/../handlers/ShowDependenciesProfilesHandler",
            customize: {},
            options: [],
            positionals: [{
                    name: constants_1.Constants.PROFILE_NAME_OPTION,
                    description: utilities_1.TextUtils.formatMessage(messages_1.deleteProfileNameDesc.message, {
                        type: this.mProfileType,
                        typeOption: profiles_1.ProfileUtils.getProfileOption(this.mProfileType)
                    }),
                    type: "string",
                    required: true,
                }]
        };
        profileCommand.customize[profiles_1.ProfilesConstants.PROFILES_COMMAND_TYPE_KEY] = this.mProfileType;
        return profileCommand;
    }
}
exports.ProfilesShowDependenciesCommandBuilder = ProfilesShowDependenciesCommandBuilder;
