// @zowe/imperative/lib/imperative/src/profiles/builders/ProfilesCreateCommandBuilder.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
Object.defineProperty(exports, "__esModule", { value: true });
const ProfilesCommandBuilder_1 = require("./ProfilesCommandBuilder");
const messages_1 = require("../../../../messages");
const constants_1 = require("../../../../constants");
const utilities_1 = require("../../../../utilities");
const profiles_1 = require("../../../../profiles");
/**
 * Used to build profile create command definitions.
 * Used automatically if you allow the "profiles" command group to be generated
 */
class ProfilesCreateCommandBuilder extends ProfilesCommandBuilder_1.ProfilesCommandBuilder {
    /**
     * Construct the builder based on the schema.
     * @param {string} profileType - the type name for the profile
     * @param {Logger} logger - logger instance to use for the builder class
     * @param {ICommandProfileTypeConfiguration} profileConfig: Imperative profile configuration for this type of profile
     */
    constructor(profileType, logger, profileConfig) {
        super(profileType, logger, profileConfig);
    }
    /**
     * Gets the "action" that this command builder is building.
     * @return {string}: The "create" action string
     */
    getAction() {
        return constants_1.Constants.CREATE_ACTION;
    }
    /**
     * Build the full command - includes action group and object command.
     * @return {ICommandDefinition}: The command definition.
     */
    buildFull() {
        return this.buildProfileSegmentFromSchema();
    }
    /**
     * Builds only the "profile" segment from the profile schema.
     * @return {ICommandDefinition}
     */
    buildProfileSegmentFromSchema() {
        const profileCommand = {
            name: this.mProfileType + "-profile",
            aliases: [this.mProfileType],
            summary: utilities_1.TextUtils.formatMessage(messages_1.createProfileCommandDesc.message, { type: this.mProfileType }),
            description: this.mSchema.description,
            type: "command",
            handler: __dirname + "/../handlers/CreateProfilesHandler",
            customize: {},
            positionals: [{
                    name: constants_1.Constants.PROFILE_NAME_OPTION,
                    description: utilities_1.TextUtils.formatMessage(messages_1.createProfileOptionDesc.message, {
                        type: this.mProfileType,
                        typeOption: profiles_1.ProfileUtils.getProfileOption(this.mProfileType)
                    }),
                    type: "string",
                    required: true,
                }],
            options: this.buildOptionsFromProfileSchema(this.mSchema.properties, [])
        };
        profileCommand.customize[profiles_1.ProfilesConstants.PROFILES_COMMAND_TYPE_KEY] = this.mProfileType;
        profileCommand.options.push({
            name: constants_1.Constants.OVERWRITE_OPTION, aliases: ["ow"],
            description: utilities_1.TextUtils.formatMessage(messages_1.createProfileOptionOverwriteDesc.message, { type: this.mProfileType }),
            type: "boolean"
        });
        if (this.mProfileConfig.createProfileExamples != null) {
            profileCommand.examples = this.mProfileConfig.createProfileExamples;
        }
        return profileCommand;
    }
}
exports.ProfilesCreateCommandBuilder = ProfilesCreateCommandBuilder;
