// @zowe/imperative/lib/imperative/src/plugins/utilities/NpmFunctions.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
Object.defineProperty(exports, "__esModule", { value: true });
const PMFConstants_1 = require("./PMFConstants");
const path = require("path");
const child_process_1 = require("child_process");
const npmCmd = cmdToRun();
/**
 * Common function that requires npm and node.exe if not found just returns npm command as a string.
 *
 * @return {string} command with node.exe and npm paths or 'npm'
 *
 */
function cmdToRun() {
    let command;
    try {
        const npmExecPath = path.join(require.resolve("npm"), "../..");
        const nodeExecPath = process.execPath;
        command = `"${nodeExecPath}" "${npmExecPath}"`;
    }
    catch (err) {
        command = "npm";
    }
    return command;
}
exports.cmdToRun = cmdToRun;
/**
 * Common function that installs a npm package using the local npm cli.
 * @param {string} prefix Path where to install npm the npm package.
 *
 * @param {string} registry The npm registry to install from.
 *
 * @param {string} npmPackage The name of package to install.
 *
 * @return {string} command response
 *
 */
function installPackages(prefix, registry, npmPackage) {
    const pipe = ["pipe", "pipe", process.stderr];
    try {
        const execOutput = child_process_1.execSync(`${npmCmd} install "${npmPackage}" --prefix "${prefix}" ` +
            `-g --registry "${registry}"`, {
            cwd: PMFConstants_1.PMFConstants.instance.PMF_ROOT,
            stdio: pipe
        });
        return execOutput.toString();
    }
    catch (err) {
        throw (err.message);
    }
}
exports.installPackages = installPackages;
/**
 * Get the registry to install to.
 *
 * @return {string}
 */
function getRegistry() {
    try {
        const execOutput = child_process_1.execSync(`${npmCmd} config get registry`);
        return execOutput.toString();
    }
    catch (err) {
        throw (err.message);
    }
}
exports.getRegistry = getRegistry;
/**
 * NPM login to be able to install from secure registry
 * @param {string} registry The npm registry to install from.
 */
function npmLogin(registry) {
    try {
        child_process_1.execSync(`${npmCmd} adduser --registry ${registry} ` +
            `--always-auth --auth-type=legacy`, { stdio: [0, 1, 2] });
    }
    catch (err) {
        throw (err.message);
    }
}
exports.npmLogin = npmLogin;
