// @zowe/cli/lib/workflows/src/cli/delete/Delete.common.handler.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const imperative_1 = require("@zowe/imperative");
const Delete_1 = require("../../api/Delete");
const ZosmfBaseHandler_1 = require("../../../../zosmf/src/ZosmfBaseHandler");
const __1 = require("../../..");
/**
 * Common handler to delete a workflow instance in z/OSMF.
 * This is not something that is intended to be used outside of this npm package.
 */
class DeleteCommonHandler extends ZosmfBaseHandler_1.ZosmfBaseHandler {
    /**
     * Command handler process - invoked by the command processor to handle the "zos-workflows delete"
     * @param {IHandlerParameters} params - Command handler parameters
     * @returns {Promise<void>} - Fulfilled when the command completes successfully OR rejected with imperative error
     * @memberof DeleteCommonHandler
     */
    processCmd(params) {
        return __awaiter(this, void 0, void 0, function* () {
            let error;
            let resp;
            let getWfKey;
            this.arguments = params.arguments;
            let sourceType;
            if (this.arguments.workflowKey) {
                sourceType = "workflowKey";
            }
            else if (this.arguments.workflowName) {
                sourceType = "workflowName";
            }
            switch (sourceType) {
                case "workflowKey":
                    try {
                        yield Delete_1.DeleteWorkflow.deleteWorkflow(this.mSession, this.arguments.workflowKey);
                    }
                    catch (err) {
                        error = "Delete workflow: " + err;
                        throw error;
                    }
                    params.response.data.setObj("Deleted.");
                    params.response.console.log("Workflow deleted.");
                    break;
                case "workflowName":
                    getWfKey = yield __1.ListWorkflows.listWorkflows(this.mSession, undefined, this.arguments.workflowName);
                    if (getWfKey === null || getWfKey.workflows.length === 0) {
                        throw new imperative_1.ImperativeError({
                            msg: `No workflows match the provided workflow name.`,
                            additionalDetails: JSON.stringify(params)
                        });
                    }
                    const successWfs = [];
                    const failedWfs = [];
                    for (const element of getWfKey.workflows) {
                        try {
                            resp = yield Delete_1.DeleteWorkflow.deleteWorkflow(this.mSession, element.workflowKey);
                            successWfs.push(element);
                        }
                        catch (err) {
                            failedWfs.push(element);
                        }
                    }
                    params.response.data.setObj("Deleted.");
                    if (getWfKey.workflows.length > 0) {
                        params.response.console.log("Successfully deleted workflow(s): ");
                        params.response.format.output({
                            fields: ["workflowName", "workflowKey"],
                            output: successWfs,
                            format: "table",
                            header: true,
                        });
                    }
                    if (failedWfs.length > 0) {
                        params.response.console.log("\nFailed to delete Workflow(s): ");
                        params.response.format.output({
                            fields: ["workflowName", "workflowKey"],
                            output: failedWfs,
                            format: "table",
                            header: true,
                        });
                        throw new imperative_1.ImperativeError({
                            msg: `Some workflows were not deleted, please check the message above.`
                        });
                    }
                    break;
                default:
                    throw new imperative_1.ImperativeError({
                        msg: `Internal create error: Unable to determine the the criteria by which to run delete workflow action. ` +
                            `Please contact support.`,
                        additionalDetails: JSON.stringify(params)
                    });
            }
        });
    }
}
exports.default = DeleteCommonHandler;
