// @zowe/imperative/lib/imperative/src/config/cmd/reset/reset.definition.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
/**
 * Definition of the list command.
 * @type {ICommandDefinition}
 */
exports.resetDefinition = {
    name: "reset",
    aliases: ["re"],
    type: "command",
    handler: path_1.join(__dirname, "reset.handler"),
    summary: "Reset a configuration setting to default value",
    description: "Reset a configuration setting to default value.",
    positionals: [
        {
            name: "configName",
            type: "string",
            description: "Setting name to reset",
            required: true
        },
    ],
    examples: [
        {
            options: "CredentialManager",
            description: "Reset the credential manager to default value"
        }
    ],
};
