// @zowe/cli/lib/zostso/src/cli/constants/ZosTso.constants.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.TSO_OPTION_GROUP = "TSO ADDRESS SPACE OPTIONS";
exports.TSO_OPTION_ACCOUNT = {
    name: "account",
    aliases: ["a"],
    description: "Your z/OS TSO/E accounting information.",
    type: "string",
    required: true,
    group: exports.TSO_OPTION_GROUP
};
exports.TSO_OPTION_CHAR_SET = {
    name: "character-set",
    aliases: ["cs"],
    description: "Character set for address space to convert messages and responses from UTF-8 to EBCDIC.",
    type: "string",
    defaultValue: "697",
    group: exports.TSO_OPTION_GROUP
};
exports.TSO_OPTION_CODE_PAGE = {
    name: "code-page",
    aliases: ["cp"],
    description: "Codepage value for TSO/E address space to convert messages and responses from UTF-8 to EBCDIC.",
    type: "string",
    defaultValue: "1047",
    group: exports.TSO_OPTION_GROUP
};
exports.TSO_OPTION_COLUMNS = {
    name: "columns",
    aliases: ["cols"],
    description: "The number of columns on a screen.",
    type: "number",
    defaultValue: 80,
    group: exports.TSO_OPTION_GROUP
};
exports.TSO_OPTION_LOGON_PROCEDURE = {
    name: "logon-procedure",
    aliases: ["l"],
    description: "The logon procedure to use when creating TSO procedures on your behalf.",
    type: "string",
    defaultValue: "IZUFPROC",
    group: exports.TSO_OPTION_GROUP
};
exports.TSO_OPTION_REGION_SIZE = {
    name: "region-size",
    aliases: ["rs"],
    description: "Region size for the TSO/E address space.",
    type: "number",
    defaultValue: 4096,
    group: exports.TSO_OPTION_GROUP
};
exports.TSO_OPTION_ROWS = {
    name: "rows",
    description: "The number of rows on a screen.",
    type: "number",
    defaultValue: 24,
    group: exports.TSO_OPTION_GROUP
};
exports.TSO_PROFILE_OPTIONS = [
    exports.TSO_OPTION_ACCOUNT,
    exports.TSO_OPTION_CHAR_SET,
    exports.TSO_OPTION_CODE_PAGE,
    exports.TSO_OPTION_COLUMNS,
    exports.TSO_OPTION_LOGON_PROCEDURE,
    exports.TSO_OPTION_REGION_SIZE,
    exports.TSO_OPTION_ROWS
];
