// @zowe/cli/lib/zosfiles/src/cli/create/classicPds/ClassicPDS.definition.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
Object.defineProperty(exports, "__esModule", { value: true });
const create_1 = require("../../../api/methods/create");
const Create_options_1 = require("../Create.options");
// Does not use the import in anticipation of some internationalization work to be done later.
const strings = require("../../-strings-/en").default.CREATE;
/**
 * Create Classic PDS command definition containing its description, examples and/or options
 * @type {ICommandDefinition}
 */
exports.ClassicPDSDefinition = {
    name: "data-set-classic",
    aliases: ["classic"],
    description: strings.ACTIONS.DATA_SET_CLASSIC.DESCRIPTION,
    type: "command",
    handler: __dirname + "/ClassicPDS.handler",
    profile: {
        optional: ["zosmf"],
    },
    positionals: [
        {
            name: "dataSetName",
            type: "string",
            description: strings.POSITIONALS.DATASETNAME,
            required: true,
        },
    ],
    options: [
        Object.assign({}, Create_options_1.ZosFilesCreateExtraOptions.size, { defaultValue: create_1.CreateDefaults.DATA_SET.CLASSIC.primary + create_1.CreateDefaults.DATA_SET.CLASSIC.alcunit }),
        Create_options_1.ZosFilesCreateOptions.volser,
        Object.assign({}, Create_options_1.ZosFilesCreateOptions.secondary, { defaultValue: create_1.CreateDefaults.DATA_SET.CLASSIC.secondary }),
        Object.assign({}, Create_options_1.ZosFilesCreateOptions.dirblk, { defaultValue: create_1.CreateDefaults.DATA_SET.CLASSIC.dirblk }),
        Object.assign({}, Create_options_1.ZosFilesCreateOptions.recfm, { defaultValue: create_1.CreateDefaults.DATA_SET.CLASSIC.recfm }),
        Object.assign({}, Create_options_1.ZosFilesCreateOptions.blksize, { defaultValue: create_1.CreateDefaults.DATA_SET.CLASSIC.blksize }),
        Object.assign({}, Create_options_1.ZosFilesCreateOptions.lrecl, { defaultValue: create_1.CreateDefaults.DATA_SET.CLASSIC.lrecl }),
        Create_options_1.ZosFilesCreateOptions.storeclass,
        Create_options_1.ZosFilesCreateOptions.mgntclass,
        Create_options_1.ZosFilesCreateOptions.dataclass,
        Create_options_1.ZosFilesCreateOptions.unit,
        Create_options_1.ZosFilesCreateOptions.dsntype,
        Create_options_1.ZosFilesCreateExtraOptions.showAttributes,
    ].sort((a, b) => a.name.localeCompare(b.name)),
    examples: [
        {
            description: strings.ACTIONS.DATA_SET_CLASSIC.EXAMPLES.EX1,
            options: "NEW.CLASSIC.DATASET",
        }
    ]
};
