// @zowe/cli/lib/zosfiles/src/cli/ZosFiles.definition.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
const Create_definition_1 = require("./create/Create.definition");
const Delete_definition_1 = require("./delete/Delete.definition");
const Invoke_definition_1 = require("./invoke/Invoke.definition");
const Download_definition_1 = require("./download/Download.definition");
const List_definition_1 = require("./list/List.definition");
const Upload_definition_1 = require("./upload/Upload.definition");
const zosmf_1 = require("../../../zosmf");
/**
 * This object defines the top level command group for zosfiles. This is not
 * something that is intended to be used outside of this npm package.
 *
 * @private
 */
const definition = {
    name: "zos-files",
    aliases: ["files"],
    type: "group",
    summary: "Manage z/OS data sets",
    description: "Manage z/OS data sets, create data sets, and more",
    children: [
        Create_definition_1.CreateDefinition,
        Delete_definition_1.DeleteDefinition,
        Invoke_definition_1.InvokeDefinition,
        Download_definition_1.DownloadDefinition,
        List_definition_1.ListDefinition,
        Upload_definition_1.UploadDefinition
    ],
    passOn: [
        {
            property: "options",
            value: zosmf_1.ZosmfSession.ZOSMF_CONNECTION_OPTIONS,
            merge: true,
            ignoreNodes: [
                { type: "group" }
            ]
        }
    ]
};
module.exports = definition;
