// @zowe/cli/lib/zosfiles/src/cli/upload/dtu/DirToUSSDir.handler.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const imperative_1 = require("@zowe/imperative");
const upload_1 = require("../../../api/methods/upload");
const api_1 = require("../../../api");
const ZosFilesBase_handler_1 = require("../../ZosFilesBase.handler");
const path = require("path");
const fs = require("fs");
/**
 * Handler to upload content from a local directory to a USS directory
 * @export
 */
class DirToUSSDirHandler extends ZosFilesBase_handler_1.ZosFilesBaseHandler {
    processWithSession(commandParameters, session) {
        return __awaiter(this, void 0, void 0, function* () {
            const status = {
                statusMessage: "Uploading all files",
                percentComplete: 0,
                stageName: imperative_1.TaskStage.IN_PROGRESS
            };
            let inputDir;
            // resolving to full path if local path passed is not absolute
            if (path.isAbsolute(commandParameters.arguments.inputDir)) {
                inputDir = commandParameters.arguments.inputDir;
            }
            else {
                inputDir = path.resolve(commandParameters.arguments.inputDir);
            }
            let response;
            const attributesFile = this.findAttributesFile(commandParameters, inputDir);
            if (attributesFile) {
                response = yield this.uploadWithAttributesFile(attributesFile, response, session, inputDir, commandParameters, status);
            }
            else {
                const filesMap = this.buildFilesMap(commandParameters);
                if (commandParameters.arguments.recursive) {
                    response = yield upload_1.Upload.dirToUSSDirRecursive(session, inputDir, commandParameters.arguments.USSDir, {
                        binary: commandParameters.arguments.binary,
                        filesMap,
                        maxConcurrentRequests: commandParameters.arguments.maxConcurrentRequests,
                        task: status
                    });
                }
                else {
                    response = yield upload_1.Upload.dirToUSSDir(session, inputDir, commandParameters.arguments.USSDir, {
                        binary: commandParameters.arguments.binary,
                        filesMap,
                        maxConcurrentRequests: commandParameters.arguments.maxConcurrentRequests,
                        task: status
                    });
                }
            }
            const formatMessage = imperative_1.TextUtils.prettyJson(response.apiResponse);
            commandParameters.response.console.log(formatMessage);
            return response;
        });
    }
    findAttributesFile(commandParameters, inputDir) {
        let attributesFile;
        if (commandParameters.arguments.attributes) {
            if (!fs.existsSync(commandParameters.arguments.attributes)) {
                throw new imperative_1.ImperativeError({ msg: imperative_1.TextUtils.formatMessage(api_1.ZosFilesMessages.attributesFileNotFound.message, { file: commandParameters.arguments.attributes }) });
            }
            attributesFile = commandParameters.arguments.attributes;
        }
        else {
            const localAttributesFile = path.join(inputDir, ".zosattributes");
            if (fs.existsSync(localAttributesFile)) {
                attributesFile = localAttributesFile;
            }
        }
        return attributesFile;
    }
    uploadWithAttributesFile(attributesFile, response, session, inputDir, commandParameters, status) {
        return __awaiter(this, void 0, void 0, function* () {
            let attributesFileContents;
            try {
                attributesFileContents = fs.readFileSync(attributesFile).toString();
            }
            catch (err) {
                throw new imperative_1.ImperativeError({ msg: imperative_1.TextUtils.formatMessage(api_1.ZosFilesMessages.errorReadingAttributesFile.message, { file: attributesFile, message: err.message }) });
            }
            let attributes;
            try {
                attributes = new api_1.ZosFilesAttributes(attributesFileContents, inputDir);
            }
            catch (err) {
                throw new imperative_1.ImperativeError({ msg: imperative_1.TextUtils.formatMessage(api_1.ZosFilesMessages.errorParsingAttributesFile.message, { file: attributesFile, message: err.message }) });
            }
            if (commandParameters.arguments.recursive) {
                response = yield upload_1.Upload.dirToUSSDirRecursive(session, inputDir, commandParameters.arguments.USSDir, {
                    attributes,
                    maxConcurrentRequests: commandParameters.arguments.maxConcurrentRequests,
                    task: status
                });
            }
            else {
                response = yield upload_1.Upload.dirToUSSDir(session, inputDir, commandParameters.arguments.USSDir, {
                    attributes,
                    maxConcurrentRequests: commandParameters.arguments.maxConcurrentRequests,
                    task: status
                });
            }
            return response;
        });
    }
    buildFilesMap(commandParameters) {
        let filesMap = null;
        // checking if binary-files or ascii-files are used, and update filesMap argument
        if (commandParameters.arguments.binaryFiles) {
            filesMap = {
                binary: true,
                fileNames: commandParameters.arguments.binaryFiles.split(",").map((fileName) => fileName.trim()),
            };
        }
        if (commandParameters.arguments.asciiFiles) {
            filesMap = {
                binary: false,
                fileNames: commandParameters.arguments.asciiFiles.split(",").map((fileName) => fileName.trim()),
            };
        }
        return filesMap;
    }
}
exports.default = DirToUSSDirHandler;
