// @zowe/cli/lib/zosfiles/src/api/methods/create/CreateDataSetType.enum.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Enumerated type that describes the different kinds of data sets that can be created
 */
var CreateDataSetTypeEnum;
(function (CreateDataSetTypeEnum) {
    CreateDataSetTypeEnum[CreateDataSetTypeEnum["DATA_SET_BINARY"] = 0] = "DATA_SET_BINARY";
    CreateDataSetTypeEnum[CreateDataSetTypeEnum["DATA_SET_C"] = 1] = "DATA_SET_C";
    CreateDataSetTypeEnum[CreateDataSetTypeEnum["DATA_SET_CLASSIC"] = 2] = "DATA_SET_CLASSIC";
    CreateDataSetTypeEnum[CreateDataSetTypeEnum["DATA_SET_PARTITIONED"] = 3] = "DATA_SET_PARTITIONED";
    CreateDataSetTypeEnum[CreateDataSetTypeEnum["DATA_SET_SEQUENTIAL"] = 4] = "DATA_SET_SEQUENTIAL";
})(CreateDataSetTypeEnum = exports.CreateDataSetTypeEnum || (exports.CreateDataSetTypeEnum = {}));
