// @zowe/cli/lib/zosjobs/src/cli/ZosJobs.definition.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
Object.defineProperty(exports, "__esModule", { value: true });
const Submit_definition_1 = require("./submit/Submit.definition");
const View_definition_1 = require("./view/View.definition");
const List_definition_1 = require("./list/List.definition");
const Cancel_definition_1 = require("./cancel/Cancel.definition");
const Delete_definition_1 = require("./delete/Delete.definition");
const Download_definition_1 = require("./download/Download.definition");
const zosmf_1 = require("../../../zosmf");
exports.definition = {
    name: "zos-jobs",
    aliases: ["jobs"],
    type: "group",
    summary: "Manage z/OS jobs",
    description: "Manage z/OS jobs.",
    children: [
        Submit_definition_1.SubmitDefinition,
        Download_definition_1.DownloadDefinition,
        View_definition_1.ViewDefinition,
        List_definition_1.ListDefinition,
        Delete_definition_1.DeleteDefinition,
        Cancel_definition_1.CancelDefinition,
    ],
    passOn: [
        {
            property: "options",
            value: zosmf_1.ZosmfSession.ZOSMF_CONNECTION_OPTIONS,
            merge: true,
            ignoreNodes: [
                { type: "group" }
            ]
        }
    ]
};
module.exports = exports.definition;
