// @zowe/cli/lib/provisioning/src/api/ProvisioningConstants.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
Object.defineProperty(exports, "__esModule", { value: true });
const imperative_1 = require("@zowe/imperative");
/**
 * Constants for provisioning related info.
 * @export
 * @class ProvisioningConstants
 */
class ProvisioningConstants {
}
/**
 * URI base for provisioning API.
 * @static
 * @type {string}
 * @memberof ProvisioningConstants
 */
ProvisioningConstants.RESOURCE = "/zosmf/provisioning/rest";
/**
 * URI base for provisioned instances API.
 * @static
 * @type {string}
 * @memberof ProvisioningConstants
 */
ProvisioningConstants.INSTANCES_RESOURCE = "scr";
/**
 * URI base for published templates API.
 * @static
 * @type {string}
 * @memberof ProvisioningConstants
 */
ProvisioningConstants.TEMPLATES_RESOURCES = "psc";
/**
 * URI base for performed actions on a provisioned instance.
 * @static
 * @type {string}
 * @memberof ProvisioningConstants
 */
ProvisioningConstants.ACTIONS_RESOURCES = "actions";
/**
 * URI base for provisioned instance variables.
 * @static
 * @type {string}
 * @memberof ProvisioningConstants
 */
ProvisioningConstants.VARIABLES_RESOURCE = "variables";
/**
 * Parameter for filtering by type.
 * @static
 * @type {string}
 * @memberof ProvisioningConstants
 */
ProvisioningConstants.RESOURCE_TYPE = "type";
/**
 * Parameter for filtering by external name.
 * @static
 * @type {string}
 * @memberof ProvisioningConstants
 */
ProvisioningConstants.RESOURCE_EXTERNAL_NAME = "external-name";
/**
 * URI base for provisioning a published template.
 * @static
 * @type {string}
 * @memberof ProvisioningConstants
 */
ProvisioningConstants.RESOURCE_PROVISION_RUN = "run";
/**
 * Version of the z/OSMF software services template service
 * @static
 * @type {string}
 * @memberOf ProvisioningConstants
 */
ProvisioningConstants.ZOSMF_VERSION = "1.0";
exports.ProvisioningConstants = ProvisioningConstants;
/**
 * Error message that no session provided.
 * @static
 * @type {IMessageDefinition}
 * @memberof ProvisioningConstants
 */
exports.noSessionProvisioning = {
    message: imperative_1.apiErrorHeader + `No session was supplied.`
};
/**
 * Error message that no z/OSMF version parameter string was supplied.
 * @static
 * @type {IMessageDefinition}
 * @memberof ProvisioningConstants
 */
exports.nozOSMFVersion = {
    message: imperative_1.apiErrorHeader + `No z/OSMF version parameter was supplied.`
};
/**
 * Error message that no template name parameter string was supplied.
 * @static
 * @type {IMessageDefinition}
 * @memberof ProvisioningConstants
 */
exports.noTemplateName = {
    message: imperative_1.apiErrorHeader + `No template name parameter was supplied.`
};
/**
 * No account number was provided error message.
 * @static
 * @type {IMessageDefinition}
 * @memberof ProvisioningConstants
 */
exports.noAccountInfo = {
    message: imperative_1.apiErrorHeader + `No account number was supplied.`
};
/**
 * Error message that no instance-id parameter string was supplied.
 * @static
 * @type {IMessageDefinition}
 * @memberof ProvisioningConstants
 */
exports.noInstanceId = {
    message: imperative_1.apiErrorHeader + `No instance-id parameter was supplied.`
};
/**
 * Error message that no template name parameter string was supplied.
 * @static
 * @type {IMessageDefinition}
 * @memberof ProvisioningConstants
 */
exports.noActionName = {
    message: imperative_1.apiErrorHeader + `No action name parameter was supplied.`
};
